### PR TITLE
Annotate potential -Wunused-function violations in err.h

### LINKED ISSUE #12792: Potentially unused static functions in err.h starting in 3.0.0
I noticed this while experimenting with integration against a fairly old version of bind9.

```
+++ Compiling: src/contrib/bind9/lib/dns/gssapi_link.c ? bedrock/private/x86_64.debug/uliba/LIBDNS.A/gssapi_link.o
In file included from src/contrib/bind9/lib/dns/gssapi_link.c:33:
In file included from src/contrib/bind9/lib/dns/dst_internal.h:57:
bedrock/import/x86_64.debug/uliba/headers/openssl/err.h:224:24: error: unused function 'ERR_GET_LIB' [-Werror,-Wunused-function]
static ossl_inline int ERR_GET_LIB(unsigned long errcode)
                       ^
bedrock/import/x86_64.debug/uliba/headers/openssl/err.h:231:24: error: unused function 'ERR_GET_FUNC' [-Werror,-Wunused-function]
static ossl_inline int ERR_GET_FUNC(unsigned long errcode ossl_unused)
                       ^
bedrock/import/x86_64.debug/uliba/headers/openssl/err.h:243:24: error: unused function 'ERR_GET_REASON' [-Werror,-Wunused-function]
static ossl_inline int ERR_GET_REASON(unsigned long errcode)
                       ^
bedrock/import/x86_64.debug/uliba/headers/openssl/err.h:250:24: error: unused function 'ERR_FATAL_ERROR' [-Werror,-Wunused-function]
static ossl_inline int ERR_FATAL_ERROR(unsigned long errcode)
                       ^
4 errors generated.
```

Ok, so in 3.0.0 instead of being macros (as they previously were), now they're static inline functions.  All fine, I suppose, but it seems like the issue is that as a freestanding header there's no guarantee that some of these functions are actually used.

It seems like the "fix" here would be to slap ossl_unused into the method signature. This would as specifically applies to these 4 functions probably be something that could probably just be pushed under "trivial" CLA. However, when I tried to generalize this I had a little pause and I wanted to open an issue to allow for discussion or for someone to attack this as more of a systemic thing.

It seems to me that any static function in a public header should probably be annotated with ossl_unused. Perhaps it's unnecessary for cases where the function is used elsewhere in the same header (ERR_GET_RFLAGS(), for example), but it seems like as a matter of general policy it would reduce potential refactoring headaches. Others may disagree. 

Should only static ossl_inline functions be considered?

Should internal headers be considered? My original use case was compiling an OpenSSL consumer with -Wunused-function, but it may be desirable to compile OpenSSL itself with that flag. Presumably that basically works today, but I can imagine situations easily emerging with the right combination of #defines wherein some implementation file pulls in an internal header that defines some static function which ends up being unused within that particular translation unit. Perhaps it can be argued that these functions should be defined only under some appropriate #ifdef & that therefore such situations developing internally would reflect a mismanagement of #define / #ifdef.

Again, my personal take would be to be pretty aggressive in deploying ossl_unused with respect to headers (internal and public) just because it seems to me that there's very little downside to it. There could be reasonable disagreement to this, however.

### PR DESCRIPTION
**Pull request by jwalch**
Fixes #12792

Supersedes #12796

### CODE REVIEW

File: Configure [hunk 1]
```diff
@@ -135,6 +135,7 @@ my @gcc_devteam_warn = qw(
     -DPEDANTIC -pedantic -Wno-long-long
     -Wall
     -Wextra
+    -Wno-unused-function
```
Comment (levitte) [inline Configure:138]: Does this actually make a tangible difference for your purposes?  As far as I can tell, it wouldn't warn us of internal static functions that are unused, and I've found that warning (turned error) useful many times over, so I would rather not have that warning turned off.

Comment (jwalch) [reply Configure]: I've removed it.

**Comment by jwalch** [comment]
UGH... Accidentally created more freaking merge commits when I meant to just squash at all... I'll clean it up with another PR I guess...

**Comment by levitte** [comment]
You should learn `git rebase` and `git push - f`, that will help you avoid having to recreate your PRs

File: Configure (hunk 1 above)
Comment (kaduk) [reply Configure]: Somehow the previous discussion on this PR had given me the impression that the idea was going to be to set `-Wunused-function` (not `-Wno-unused-function`).  Maybe I missed some messages...

**Pull request closed without merging**